### PR TITLE
chore: updated sql agent example and added tests

### PR DESCRIPTION
**Pull request by brenkao**

### CODE REVIEW

File: docs/cookbook/agents/sql_agent.md [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+We will be using SQLite but this example will work for any common SQL dialect, such as PostgreSQL, MySQL, MSSQL, and more.
+
+```python
+pip install sqlite3
+```
+
+## Setup SQL Database
+
+Replace this part with whichever SQL dialect you are using, or skip if you have a database setup already.
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:25]: nit: "setup" should be "set up" in this case I believe, could instead say "if you have already setup a database"

File: docs/cookbook/agents/sql_agent.md [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+```
+
+This will be our playground example.
+
+## Write your Database Assistant
+
+We will be creating an Agent that will take non-technical queries and translate them into SQL queries that will be executed. The first step will be to create our two tools, `_run_query` and `_execute_query` , which will be read and write operations respectively.
+
+```python
+import sqlite3
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:50]: import ordering is wrong here, should be:
```python
import sqlite3
from typing import ClassVar

from mirascope.core import openai, prompt_template
from pydantic import BaseModel, ConfigDict
```

File: docs/cookbook/agents/sql_agent.md [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+        You have access to the following tools:
+        - `_run_query`: Use the query generated to run a SELECT query.
+        - `_execute_query`: Use the query generated to execute an 
+            INSERT, UPDATE, or DELETE query.
+
+        You must use these tools to interact with the database.
+
+        MESSAGES: {self.messages}
+        USER:
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:167]: nit: put this on the same line like you do for MESSAGES

File: docs/cookbook/agents/sql_agent.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        {query}
+        """
+    )
+    async def _step(self, query: str) -> openai.OpenAIDynamicConfig:
+        return {
+            "tools": [self._run_query, self._execute_query],
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:173]: nit, comma and formatting here seem unnecessary?

File: docs/cookbook/agents/sql_agent.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                tools_and_outputs.append((tool, tool.call()))
+            else:
+                print(chunk.content, end="", flush=True)
+        if response.user_message_param:
+            self.messages.append(response.user_message_param)
+        self.messages.append(response.message_param)
+        if tools_and_outputs:
+            self.messages += response.tool_message_params(tools_and_outputs)
+            return await self._get_response("")
+        return
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:190]: unnecessary return

File: docs/cookbook/agents/sql_agent.md [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+            if tool:
+                tools_and_outputs.append((tool, tool.call()))
+            else:
+                print(chunk.content, end="", flush=True)
+        if response.user_message_param:
+            self.messages.append(response.user_message_param)
+        self.messages.append(response.message_param)
+        if tools_and_outputs:
+            self.messages += response.tool_message_params(tools_and_outputs)
+            return await self._get_response("")
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:189]: unnecessary return

File: docs/cookbook/agents/sql_agent.md [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    async for _, tool in response:
+        query = tool.args.get("query", "") if tool else ""
+        assert query == "DELETE FROM ReadingList WHERE title = 'Gone with the Wind';"
+```
+
+The golden dataset serves as our test fixtures, allowing us to verify basic text-to-SQL conversions and ensure our Agent maintains this functionality.
+
+As we expand our tools or enhance our prompt, we should consistently implement additional prompt regression tests to maintain quality and functionality.
+
+
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:325]: nit: seems like a bunch of newlines here that aren't needed?

File: docs/cookbook/agents/sql_agent.md [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+async def test_delete_query(delete_query: str, mock_librarian: Librarian):
+    response = await mock_librarian._step(delete_query)
+    async for _, tool in response:
+        query = tool.args.get("query", "") if tool else ""
+        assert query == "DELETE FROM ReadingList WHERE title = 'Gone with the Wind';"
+```
+
+The golden dataset serves as our test fixtures, allowing us to verify basic text-to-SQL conversions and ensure our Agent maintains this functionality.
+
+As we expand our tools or enhance our prompt, we should consistently implement additional prompt regression tests to maintain quality and functionality.
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:323]: should we add a warning or tip or something talking about the fact that SQL can get very complex and this is a fairly simple example i.e. make sure to add examples that are really good / heavily test etc?

I feel like even a short blurb about this would go a long way

File: examples/cookbook/agents/sql_agent/agent.py
Comment (willbakst) [inline examples/cookbook/agents/sql_agent/agent.py:1]: please apply changes mentioned in markdown review to this code as well

File: docs/cookbook/agents/sql_agent.md (hunk 6 above)
Comment (brenkao) [reply docs/cookbook/agents/sql_agent.md:188]: how is this one unnecessary?

File: docs/cookbook/agents/sql_agent.md [hunk 9]
```diff
@@ -164,14 +162,11 @@ class Librarian(BaseModel):
...
+        USER: {query}
         """
     )
     async def _step(self, query: str) -> openai.OpenAIDynamicConfig:
-        return {
-            "tools": [self._run_query, self._execute_query],
-        }
+        return {"tools": [self._run_query, self._execute_query]}
         
     async def _get_response(self, question: str):
```
Comment (willbakst) [inline docs/cookbook/agents/sql_agent.md:171]: maybe `_get_response` is not the right naming for this function since it doesn't return anything.

I think a better structure would be `run` calls `_step` that calls `_call`

File: docs/cookbook/evals/index.md
Comment (willbakst) [inline docs/cookbook/evals/index.md:1]: # Evaluation Cookbook

This collection of recipes focuses on techniques and strategies for evaluating and improving the performance of your LLM-powered applications (whether or not you've built them with Mirascope). Proper evaluation is crucial for ensuring the reliability, accuracy, and effectiveness of AI systems, especially when working with large language models.

## Why Evaluate?

Evaluating your LLM applications is essential for several reasons:

1. **Quality Assurance**: Ensure your models are producing accurate and relevant outputs.
2. **Performance Monitoring**: Track how your models perform over time and across different inputs.
3. **Continuous Improvement**: Gather insights to refine your prompts, fine-tune your models, or adjust your system architecture.
5. **User Experience**: Guarantee a consistent and high-quality experience for your end-users.

## What You'll Find Here

Each recipe in this section has a sister recipe containing the original recipe implementing the system we will evaluate. These recipes will:

- Provides practical code examples using Mirascope
- Explains key concepts and best practices in LLM evaluation
- Offers tips for adapting the evaluation methods to your specific use case

## Recipes In This Section

Explore the following recipes to learn how to effectively evaluate your LLM applications with Mirascope:

- [Evaluating an SQL Agent](./evaluating_sql_agent.md): Learn how to assess and improve the quality of SQL queries generated by LLMs.

## How to Use This Cookbook

1. Review the evaluation recipes to find techniques relevant to your project.
2. Implement the evaluation methods alongside your main application logic.
3. Use the insights gained from evaluations to iteratively improve your LLM applications.
4. Consider combining multiple evaluation techniques for a more comprehensive assessment.

## Best Practices

While it may be tempting to use existing generalized evaluation tooling, this often (if not always) results in a failure to properly evaluate the quality of your application.

When implementing your evals, remember to:

- **Clearly Define Your Criteria**: The criteria by which you measure success and quality are the cornerstone of any good evaluation system. This is the first and most important step.
- **Establish Baselines**: Create a set of benchmark tests to track improvements (or degradation) over time.
- **Use Diverse Test Sets**: Push your evals to cover a wide range of inputs and edge cases. 
- **Human-in-the-loop**:  Iterate on your evals with thoughtful, nuanced human reviews to ensure the metrics you're tracking remain true to your goals.
- **Automate Where Possible**: Integrate evaluation into your CI/CD pipeline for continuous monitoring.

**Review by willbakst — APPROVED** [review]
